active/0000-lazy-load-overview.md

```diff
@@ -0,0 +1,185 @@
+All of the code is removed to focus on the lazy-loading parts.
+
+```
+// app/components/z-table.js
+export default Ember.Component.extend({
+  lazyLoader: Ember.inject.service()
+  init() {
+    // This call is idempotent and the path will be mapped to a finger printed veresion
+  	this.get('lazyLoader').load('hands-on-table.js');
+  }
+});
+
+// app/components/z-table.hbs
+{{#if get(lazyLoader.loadedLibraries "hands-on-table.js"}}
+  {{!-- display other functionality that depends on the lazy-loaded lib --}}
+{{/if}}
+
+// ember-cli-build.js
+let app = new EmberApp(defaults, {});
+app.import('bower_components/hands-on-table/dist/hands-on-table.js', {outputFile: 'hands-on-table.js');
+```
+
+
+## Lazy Loading from routes
+
+For app code, we likely want an integration with the router, which can rely on this service to do the actual loading.
+
+When splitting the app's JS and CSS into multiple assets related to different sectionsof an application, we want to load the code dynamically. This today poses additional challenges because the handler (route) needs to exist to initiate a
+transition, but that route class doesn't exist until we load the new code. Also link-to depends on the route information
+in order to `serialize` the model information to create the URL. That means that even before attempting to
+transition, the route needs to be available. The latter problem was solved with the
+[serialize changes](https://github.com/emberjs/rfcs/pull/120). To add the ability to load code and lazily load routes,
+we need a new hook in Ember's router or router.js, ongoing work for this is being done at the moment to make the
+`getHandler` function asynchronous. Once that is done, we can call the loading-service similar to what we do in
+components and resolve with the name of the asset to be loaded based on route metadata. The following code is
+just an example and assumes we have the route metadata and the loadingService available.
+A final implementation of how this hook will work is out of the scope of this RFC.
+
+```
+Router.getHandler = function (routeName) {
+  let assetName = routerMetadata.getAssetFor(routeName);
+  if (lazyLoaderService.loadedLibraries["assetName"]) {
+    return this._super(...arguments);
+  }
+  return lazyLoadingService.load(assetName).then(()=> {
+    this._super(...arguments);
+  });
+};
+```
+
+### Route metadata
+
+The code above assumes that we have the routerMetadata available, the definition of this needs further discussion 
+outside of the scope of this RFC. For the purpose of this overview we simply assume that we have a way to map between routeNames and assetNames.
+
+## Engines
+
+Engines can simply rely on all of the parts above and provide nice conventions.
+
+## Dependencies and concatenation
+
+Sometimes an asset has a dependency on another asset. For example,
+[ember-cli-bundle-loader](https://github.com/MiguelMadero/ember-cli-bundle-loader/blob/master/tests/dummy/config/bundles.js)
```

Comment:
I'll update the RFC to make that clear and remove the references to that project. 

What I'm referring to here is concatenating two assets. While it might make sense to logically think of, for example, insurance and  commute as two separate "engines", we might want to concatenate them into a "benefits bundle" if we know that they will always be used together or they have dependencies. Do you see any risk in doing this?